packages/server/test/unit/cloud/api/get_app_studio_spec.ts

```diff
@@ -0,0 +1,336 @@
+      expect(studioManagerSetupStub).to.be.calledWithMatch({
+        script: 'console.log("studio script")',
+        studioPath: '/tmp/cypress/studio',
+        studioHash: undefined,
+      })
+    })
+  })
+
+  describe('CYPRESS_LOCAL_STUDIO_PATH not set', () => {
+    it('downloads the studio bundle and extracts it', async () => {
+      const readStream = Readable.from('console.log("studio script")')
+
+      crossFetchStub.resolves({
+        body: readStream,
+        headers: {
+          get: (header) => {
+            if (header === 'x-cypress-signature') {
+              return '159'
+            }
+          },
+        },
+      })
+
+      let writeResult = ''
+
+      const writeStream = new Writable({
+        write: (chunk, encoding, callback) => {
+          writeResult += chunk.toString()
+          callback()
+        },
+      })
+
+      createWriteStreamStub.returns(writeStream)
+
+      verifySignatureFromFileStub.resolves(true)
+
+      createReadStreamStub.returns(Readable.from('tar contents'))
+
+      const projectId = '12345'
+
+      await getAppStudio(projectId)
+
+      expect(rmStub).to.be.calledWith('/tmp/cypress/studio')
+      expect(ensureStub).to.be.calledWith('/tmp/cypress/studio')
+
+      expect(crossFetchStub).to.be.calledWith('http://localhost:1234/studio/bundle/current.tgz', {
+        agent: sinon.match.any,
+        method: 'GET',
+        headers: {
+          'x-route-version': '1',
+          'x-cypress-signature': '1',
+          'x-cypress-project-slug': '12345',
+          'x-cypress-studio-mount-version': '1',
+          'x-os-name': 'linux',
+          'x-cypress-version': '1.2.3',
+        },
+        encrypt: 'signed',
+      })
+
+      expect(writeResult).to.eq('console.log("studio script")')
+
+      expect(verifySignatureFromFileStub).to.be.calledWith('/tmp/cypress/studio/bundle.tar', '159')
+
+      expect(extractStub).to.be.calledWith({
+        file: '/tmp/cypress/studio/bundle.tar',
+        cwd: '/tmp/cypress/studio',
+      })
+
+      expect(readFileStub).to.be.calledWith('/tmp/cypress/studio/server/index.js', 'utf8')
+
+      expect(studioManagerSetupStub).to.be.calledWithMatch({
+        script: 'console.log("studio script")',
+        studioPath: '/tmp/cypress/studio',
+        studioHash: 'V8T1PKuSTK1h9gr-1Z2Wtx__bxTpCXWRZ57sKmPVTSs',
+      })
+    })
+
+    it('downloads the studio bundle and extracts it after 1 fetch failure', async () => {
+      const readStream = Readable.from('console.log("studio script")')
+
+      crossFetchStub.onFirstCall().rejects(new HttpError('Failed to fetch', 'url', 502, 'Bad Gateway', 'Bad Gateway', sinon.stub()))
+      crossFetchStub.onSecondCall().resolves({
+        body: readStream,
+        headers: {
+          get: (header) => {
+            if (header === 'x-cypress-signature') {
+              return '159'
+            }
+          },
+        },
+      })
+
+      let writeResult = ''
+
+      const writeStream = new Writable({
```

Comment:
are we able to refactor some of this to a beforeEach to help make the tests a bit easier to read?

Reply:
might actually make it worse to read if you do that considering the conditions change 😂

Reply:
I did a little refactoring: [`634109a` (#31033)](https://github.com/cypress-io/cypress/pull/31033/commits/634109a70223aa69af77b1fa5415ce3891556128)